substrate/runtime/contract/src/gas.rs

```diff
@@ -137,13 +155,21 @@ pub fn buy_gas<T: Trait>(
 	}
 	<staking::Module<T>>::set_free_balance(transactor, b - cost);
 	Ok(GasMeter {
+		limit: gas_limit,
 		gas_left: gas_limit,
 		gas_price,
 	})
 }
 
 /// Refund the unused gas.
 pub fn refund_unused_gas<T: Trait>(transactor: &T::AccountId, gas_meter: GasMeter<T>) {
+	// Increase total spent gas.
+	// This cannot overflow, since `gas_spent` is never greater than `block_gas_limit`, which
+	// also has T::Gas type.
+	let gas_spent = <Module<T>>::gas_spent() + gas_meter.spent();
```

Comment:
Is this correct? Are we sure to never add the same `gas_meter.spent()` twice here?

My point is that:
1. You spend some gas and store it here (so you have `B + spent` in the storage)
2. You spend some more, say `X` and you add it here, so you will have `(B + spent + (spent + X))` in the storage

I suppose it's not really a case, since we consume `gas_meter` and probably take care of proper accounting on some other layer, but wouldn't it be simpler to just store global `gas_left` in the storage?

Reply:
Yeah, that's not really the case and yes, I've tried hard to achieve the design where you have the construct like:

```
1. gas_meter = buy_gas(price)
... do stuff with gas meter
n. refund_unused_gas(gas_meter)
```

so it is an invariant that we do refund and spent gas accounting only once per transaction and the mechanism for that, as you noticed, is consuming the `GasMeter`.

Let's move our discussion why I used `gas_spent` instead of `gas_left` in [this thread](
https://github.com/paritytech/polkadot/pull/506#discussion_r208970824)!